docs/advanced/authentication.md

```diff
@@ -0,0 +1,328 @@
+```
+
+For most graphQL APIs, the auth error is communicated as an error code, however it may also be a 401 HTTP response:
+
+```js
+const didAuthError = ({ error }) => {
+  return error.graphQLErrors.some(
+    e => error.response.status === 401,
+  );
+},
+```
+
+Then `didAuthError` returns `true`, it will trigger the exchange to trigger the logic for asking for reauthentication via `getAuth`.
+
+### Configuring `getAuth` (after `didAuthError` returns `true`)
+
+If your API doesn't support any sort of token refresh, this is where you should simply log the user out.
+
+```js
+const getAuth = async ({ authState }) => {
+  if (!authState) {
+    const token = localStorage.getItem('token');
+    const refreshToken = localStorage.getItem('refreshToken');
+    if (token && refreshToken) {
+      return { token, refreshToken };
+    }
+    return null;
+  }
+
+  logout();
+
+  return null;
+}
+```
+
+Here, `logout()` should clear your persisted storage of any tokens and set the `isLoggedIn` property as described above to `false`.
+
+If you do have a token refresh endpoint, and a refresh token, you can attempt to get a new token for the user:
+
+```js
+const getAuth = async ({ authState, mutate }) => {
+  if (!authState) {
+    const token = localStorage.getItem('token');
+    const refreshToken = localStorage.getItem('refreshToken');
+    if (token && refreshToken) {
+      return { token, refreshToken };
+    }
+    return null;
+  }
+
+  const result = await mutate(refreshMutation, {
+    token: authState!.refreshToken,
+  });
+
+  if (result.data?.refreshLogin) {
+    localStorage.setItem('token', result.data.refreshLogin.token);
+    localStorage.setItem('refreshToken', result.data.refreshLogin.refreshToken);
+
+    return {
+      token: result.data.refreshLogin.token,
+      refreshToken: result.data.refreshLogin.refreshToken,
+    };
+  }
+
+  localStorage.clear();
+  logout();
+
+  return null;
+}
+```
+
+Here we use the special mutate function provided by the auth exchange to do the token refresh. If your auth is not handled via graphQL, you can
+use `fetch` in this function instead of a mutation. All other requests will be paused while `getAuth` returns, so you won't get multiple auth failures
+at the same time.
+
+### Configuring `willAuthError`
+
+`willAuthError` is an optional parameter and is run _before_ a network request is made. You can use it to trigger the logic in
+`getAuth` without the need to get a graphql error first. For example, you could use this indicate that the auth will fail because the JWT is invalid:
+
+```js
+const willAuthError = ({ authState }) => {
+  if (!authState || /* JWT is expired */) return true;
+  return false;
+}
+```
+
+## Handling Logout with the Error Exchange
+
+It is also possible to handle your logout logic in the error exchange instead of in the auth exchange. To do this, you'll need to add the
+`errorExchange` to the exchanges array, _before_ the `authExchange`. The order is very important here:
```

Comment:
```suggestion
We can also handle authentication errors in an `errorExchange` instead of the `authExchange`. To do this, we'll need to add the
`errorExchange` to the exchanges array, _before_ the `authExchange`. The order is very important here:
```